reflection/serverreflection.go

```diff
@@ -57,8 +58,10 @@ import (
 
 type serverReflectionServer struct {
 	s *grpc.Server
-	// TODO add more cache if necessary
-	serviceInfo map[string]grpc.ServiceInfo // cache for s.GetServiceInfo()
+
+	mu           sync.RWMutex
```

Comment:
That seems awkward. It would require that the reflection service be registered *last*, so that it has access to previously registered services. Any services registered with the `*grpc.Server` after it will be unknown to the reflection service. And if it is instead registered first, it will serve no information (which is actually how our internal libraries work: they register it before returning a `*grpc.Server` for calling code to add other services so that all of our internal microservices always have reflection support.)

What if I change it to use `sync.Once` instead, so the set of descriptors will be frozen after the first real method invocation? Now that I'm staring at this, I think that would make it a lot simpler and cleaner.

Reply:
Changed to use `sync.Once`.